xhistogram/test/test_chunking_hypotheses.py

```diff
@@ -0,0 +1,86 @@
+import numpy as np
+import pytest
+
+from .fixtures import example_dataarray, example_dataset
+from ..xarray import histogram
+
+pytest.importorskip("hypothesis")
+
+import hypothesis.strategies as st  # noqa
+from hypothesis import given  # noqa
+
+
+@st.composite
+def chunk_shapes(draw, n_dim=3, max_arr_len=10):
+    """Generate different chunking patterns for an N-D array of data."""
+    chunks = []
+    for n in range(n_dim):
+        shape = draw(st.integers(min_value=1, max_value=max_arr_len))
+        chunks.append(shape)
+    return tuple(chunks)
+
+
+class TestChunkingHypotheses:
+    @given(chunk_shapes(n_dim=1, max_arr_len=20))
+    def test_all_chunking_patterns_1d(self, chunks):
+
+        data = example_dataarray(shape=(20,)).chunk(chunks)
+
+        nbins_a = 8
+        bins = np.linspace(-4, 4, nbins_a + 1)
+
+        h = histogram(data, bins=[bins])
+
+        assert h.shape == (nbins_a,)
+
+        hist, _ = np.histogram(
+            data.values.ravel(),
+            bins=bins,
+        )
+
+        np.testing.assert_allclose(hist, h)
+
+    # TODO mark as slow?
+    @given(chunk_shapes(n_dim=2, max_arr_len=8))
+    def test_all_chunking_patterns_2d(self, chunks):
+
+        data_a = example_dataarray(shape=(5, 20)).chunk(chunks)
+        data_b = example_dataarray(shape=(5, 20)).chunk(chunks)
+
+        nbins_a = 8
+        nbins_b = 9
+        bins_a = np.linspace(-4, 4, nbins_a + 1)
+        bins_b = np.linspace(-4, 4, nbins_b + 1)
+
+        h = histogram(data_a, data_b, bins=[bins_a, bins_b])
+
+        assert h.shape == (nbins_a, nbins_b)
+
+        hist, _, _ = np.histogram2d(
+            data_a.values.ravel(),
+            data_b.values.ravel(),
+            bins=[bins_a, bins_b],
+        )
+
+        np.testing.assert_allclose(hist, h.values)
+
+    # TODO mark as slow?
+    @pytest.mark.parametrize("n_vars", [1, 2, 3, 4])
+    @given(chunk_shapes(n_dim=2, max_arr_len=7))
```

Comment:
What exactly do you mean? If I make the `test_all_chunking_patterns_dd_hist` accept a `dims` (or `reduce_axes`) argument then I also need a `np.histogramdd` function that can handle that generality. Is there a quick way to achieve that in the test? Possibly with `np.apply_along_axis`?

For the weights then I guess I could pass weights and allow the data and weights to have different chunking patterns - is that what you meant?

Reply:
Yeah, I suppose it's trickier to test that, since you'd need something to do N-D histograms (xhistogram) to verify the results.

I suppose you could just compare against `histogram` of the computed (NumPy) arrays, and make it purely a test of the dask functionality. If we're confident the NumPy code paths are well-tested, that seems reasonable to me.

But it was just a thought; I think the tests here are already quite good, so fine to leave it as-is too.